### PR TITLE
ColumnView

### LINKED ISSUE #147: Does anyone want a macOS/iPadOS like column view?
image

### PR DESCRIPTION
**Pull request by TyJOrtiz**
Added a columnViewBrowser.

Very barebones, only the "Layout Mode" Context menu item works.

### CODE REVIEW

File: Files/Views/LayoutModes/GenericFileBrowser.xaml [hunk 1]
```diff
@@ -98,11 +98,19 @@
                     </MenuFlyoutItem.Icon>
                 </MenuFlyoutItem>
                 <MenuFlyoutItem
-                    x:Uid="StatusBarControlGridViewSmall"
-                    Command="{x:Bind AppSettings.ToggleLayoutModeGridViewSmall}"
-                    Text="Grid View (Small)">
```
Comment (yaichenbaum) [inline Files/Views/LayoutModes/GenericFileBrowser.xaml:103]: It looks like you removed this layout option by mistake.

File: Files/Interacts/Interaction.cs [hunk 2]
```diff
@@ -613,6 +613,30 @@ private async void OpenSelectedItems(bool displayApplicationPicker)
             }
         }
 
+        internal void Prepare(string itemPath)
+        {
+            SetDirectory(itemPath);
+        }
+
+        private async void SetDirectory(string path)
+        {
```
Comment (duke7553) [inline Files/Interacts/Interaction.cs:622]: This method seems kind of redundant and probably doesn't belong in Interaction.cs. Does using/improving upon what exists in SetWorkingDirectoryAsync() solve the problem?

File: Files/Views/ModernShellPage.xaml.cs [hunk 3]
```diff
@@ -255,10 +255,11 @@ private void SidebarControl_SidebarItemInvoked(object sender, Controls.SidebarIt
                     AssociatedTabInstance = this
                 },
                 new SuppressNavigationTransitionInfo());
+            ReloadColumnView?.Invoke(this, EventArgs.Empty);
 
             NavigationToolbar.PathControlDisplayText = FilesystemViewModel.WorkingDirectory;
         }
-
+        public static event EventHandler ReloadColumnView;
```
Comment (duke7553) [inline Files/Views/ModernShellPage.xaml.cs:262]: Why would we need this event? BaseLayout-derived layout modes should load the new content upon being navigated to without the need for something like this in the shell page. 🙂

**Review by TyJOrtiz — COMMENTED** [review]
@yaichenbaum I fixed that issue in a new commit

File: Files/Views/LayoutModes/GenericFileBrowser.xaml (hunk 1 above)
Comment (TyJOrtiz) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml]: I fixed that issue

File: Files/Views/ModernShellPage.xaml.cs (hunk 3 above)
Comment (duke7553) [reply Files/Views/ModernShellPage.xaml.cs:262]: @TyJOrtiz Why didn't the OnNavigatedTo event work out? You should be able to override it iirc

Comment (TyJOrtiz) [reply Files/Views/ModernShellPage.xaml.cs:262]: Ok, I must've done something wrong there. I added the override and it works

Comment (duke7553) [reply Files/Views/ModernShellPage.xaml.cs]: Excellent!

**Comment by d2dyno1** [comment]
Bugs I've managed to spot:
- If double clicked a directory, two columns would open with the same directory
- "Continue where you left off" does not work with column view

**Comment by TyJOrtiz** [comment]
> @TyJOrtiz Good job on your PR!
> I've compiled a list of things for this PR.
> There are some smol bugs I've managed to spot 🐞 :
> (This list also includes things that need? to be done)
> 
> * If double clicked a directory, two (or more) columns would open with the same directory

Fixed

> * "Continue where you left off" does not work with column view

I have no idea how to implement that

> * Context menu is not shown for items, instead it is shown for current directory (which is incorrect) - (I can't reproduce this issue anymore. Guess it fixed it self 🤷‍♂️)
> * Rounded corners on items?

I can fix that

> * When selecting more than one item, progressbar shows up

Fixed

> * When item is selected, the context menu for this specific item is still shown when left-clicked on empty space (Not only in ColumnView)

Trying to fix this 

> * Refreshing directory in ColumnView makes the previous directory the same as the leading one

Fixed

> * Keyboard shortcuts don't work (ctrl+c,v,x,a etc.)
> * Selection Rectangle is not present

Working on it

> * Some items have a border around them:
>   image
>   .....

Should be fixed

>   and sometimes not:
>   image
> * FileSystemWatcher is not present
> * No drag & drop
> * Items have click animation whereas listview items doesn't

?

> * Selection options don't work
>   image

Working on that

File: Files/Views/LayoutModes/ColumnViewBrowser.xaml [hunk 4]
```diff
@@ -0,0 +1,2070 @@
...
+            <Setter Property="Template">
+                <Setter.Value>
+                    <ControlTemplate TargetType="ListViewItem">
+                        <Grid x:Name="ContentBorder"
+              Control.IsTemplateFocusTarget="True"
+              FocusVisualMargin="{TemplateBinding FocusVisualMargin}"
+              Background="{TemplateBinding Background}"
+              BorderBrush="{TemplateBinding BorderBrush}"
+              BorderThickness="{TemplateBinding BorderThickness}"
+              CornerRadius="0"
```
Comment (yaichenbaum) [inline Files/Views/LayoutModes/ColumnViewBrowser.xaml:46]:
```suggestion
```

File: Files/Views/LayoutModes/ColumnViewBrowser.xaml [hunk 5]
```diff
@@ -0,0 +1,2070 @@
...
```
Comment (yaichenbaum) [inline Files/Views/LayoutModes/ColumnViewBrowser.xaml:544]:
```suggestion
```

**Comment by d2dyno1** [comment]
> * "Continue where you left off" does not work with column view
@TyJOrtiz
There's a `Path` property in `ITabItem` (- newest commit) that needs to be updated when navigating to a directory.
Continue where you left off uses that property. Do you update it?

**Comment by d2dyno1** [comment]
@TyJOrtiz
There's a `Path` property in `ITabItem` (- newest commit) that needs to be updated when navigating to a directory.
Continue where you left off uses that property. Do you update it?
> > * "Continue where you left off" does not work with column view
> 
> I have no idea how to implement that

**Comment by TyJOrtiz** [comment]
> @TyJOrtiz
> There's a `Path` property in `ITabItem` that needs to be updated when navigating to a directory.
> Continue where you left off uses that property. Do you update it?
> 
> > > * "Continue where you left off" does not work with column view
> > 
> > 
> > I have no idea how to implement that

@d2dyno1 Fixed

**Comment by yaichenbaum** [comment]
@TyJOrtiz It looks lile the option to rename items isn't working in the column view layout.

**Comment by TyJOrtiz** [comment]
@yair100 I'm still working on it, I don't know how to do all that stuff yet. Like I said everything's more complex from when I first started on this so I'm just trying to acquaint myself.

**Comment by d2dyno1** [comment]
> @yaichenbaum I'm still working on it, I don't know how to do all that stuff yet. Like I said everything's more complex from when I first started on this so I'm just trying to acquaint myself.

There's a `IFilesystemHelpers` interface for this

**Comment by d2dyno1** [comment]
> * Refreshing directory in ColumnView removes all previous directories
> * Keyboard shortcuts don't work (ctrl+c,v,x,a etc.)
> * Selection Rectangle is not present
>   image
>   .....
>   and sometimes not:
>   image
> * FileSystemWatcher is not present
> * No drag & drop
> * Items should not have click animation
> * Selection options don't work
>   image
> * Phone contents aren't shown in ColumnView:
>   image
>   .
>   image

@TyJOrtiz How's progress going?

**Comment by TyJOrtiz** [comment]
> @TyJOrtiz It looks lile the option to rename items isn't working in the column view layout.

@yair100 Renaming file functionality is now there, I will need to make improvements

> Items should not have click animation

@d2dyno1 Fixed

I'll be working on this more throughout the week

**Comment by d2dyno1** [comment]
Hi @TyJOrtiz! Do you need any help with the PR?

**Comment by lucandrei** [comment]
I would love to see this live! ❤️

**Comment by yaichenbaum** [comment]
@TyJOrtiz Can we get a status update on this pull request? We would be happy to help solve the merge conflicts from `main`.

**Comment by TyJOrtiz** [comment]
I'm rewriting the whole ColumnView. I'm still trying to figure out how to make individual instances of the itemviewmodel

**Comment by mdtauk** [comment]
> I'm rewriting the whole ColumnView. I'm still trying to figure out how to make individual instances of the itemviewmodel

ColumnView probably wont be compatible with the Split Pane view.  I don't think it would make for a great UI or UX for users.

**Comment by d2dyno1** [comment]
> I'm rewriting the whole ColumnView. I'm still trying to figure out how to make individual instances of the itemviewmodel

@TyJOrtiz don't hesitate to ask for help!

**Comment by TyJOrtiz** [comment]
@d2dyno1 I figured out how to create new viewmodels so now it's full steam ahead

**Comment by lucandrei** [comment]
Let's gooo! Thank you 🙌

**Comment by bl-ue** [comment]
Any progress?

**Comment by TyJOrtiz** [comment]
I created a pull request a couple weeks ago, what I’m working on now is keyboard shortcuts and trying to fix that MTP issue.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by d2dyno1** [comment]
What if we merge ColumnView as is and enable it for debug only. And improve it with further PRs to have this PR finally out of the way?

**Comment by TyJOrtiz** [comment]
That sounds like a good idea.

Sent from my iPhone

On Feb 27, 2021, at 1:48 PM, d2dyno1 <notifications@github.com> wrote:

﻿

What if we merge ColumnView as is and enable it for debug only. And improve it with further PRs to have this PR finally out of the way?

**Comment by d2dyno1** [comment]
@TyJOrtiz please resolve conflicts

**Comment by TyJOrtiz** [comment]
Will do, I'll have it finished before the night's over

**Comment by TyJOrtiz** [comment]
I tried to see whats conflicting and it’s saying “too complex to view in browser”. I have no idea what to do

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by d2dyno1** [comment]
@TyJOrtiz in VS, click "Commit and Sync"

**Comment by d2dyno1** [comment]
if that doesn't work,
open cmd in files repo dir:

1 
```
git fetch origin master
```
2
```
git merge origin master
```
If that doesn't work, do:
 
1
```
git remote add upstream https://github.com/files-community/Files.git
```
2
```
git fetch upstream
```
3
```
git merge upstream/master
```

Conflicts will be shown in VS to resolve

**Comment by TyJOrtiz** [comment]
I finally merged the main to my fork, I have a question before I freate a pull request...

The context menu no longer shows the Windows Explorer-like context menu (for any view), now its just a basic context menu 
image

Was that intentional?

**Comment by gave92** [comment]
> The context menu no longer shows the Windows Explorer-like context menu (for any view), now its just a basic context menu
Was that intentional?

Nooope, we need to investigate why this is happening :)

**Comment by yaichenbaum** [comment]
@gave92 I have this issue with `main` as well.

**Comment by bl-ue** [comment]
Hey, why didn't this get merged? 😢

**Comment by yaichenbaum** [comment]
@bl-ue There is a new PR #4176.

**Comment by bl-ue** [comment]
Ah, I see. Thank you, @yair100!

**Pull request closed without merging**